src/i2c.rs

```diff
@@ -7,186 +7,326 @@
+            .pop_front()
+            .expect("no pending expectation for i2c::read call");
+            
+        assert_eq!(w.expected_mode, Mode::Write);
+        assert_eq!(w.expected_addr, address);
+        assert_eq!(w.expected_data, bytes);
+
         Ok(())
     }
 }
 
-impl<'a> i2c::WriteRead for Mock<'a> {
+impl i2c::WriteRead for Mock {
     type Error = MockError;
 
     fn write_read(
         &mut self,
         address: u8,
         bytes: &[u8],
-        mut buffer: &mut [u8],
+        buffer: &mut [u8],
     ) -> Result<(), Self::Error> {
-        if bytes.len() > WRITE_BUF_SIZE {
-            panic!(
-                "Write buffer is too small for this number of bytes ({} > {})",
-                bytes.len(),
-                WRITE_BUF_SIZE
-            );
+        let w = self
+            .expected
+            .lock()
+            .unwrap()
+            .pop_front()
+            .expect("no pending expectation for i2c::read call");
+            
+        assert_eq!(w.expected_mode, Mode::WriteRead);
+        assert_eq!(w.expected_addr, address);
+        assert_eq!(w.expected_data, bytes);
+
+        assert_eq!(buffer.len(), w.response_data.len(), "mismatched response length for i2c:read");
+        for i in 0..buffer.len() {
+            buffer[i] = w.response_data[i];
         }
-        self.address = Some(address);
-        self.data.read(&mut buffer)?;
-        self.buf[0..bytes.len()].copy_from_slice(bytes);
-        self.buf_bytes_written = bytes.len();
+
         Ok(())
     }
 }
 
 #[cfg(test)]
-mod tests {
+mod test {
     use super::*;
 
     use hal::blocking::i2c::{Read, Write, WriteRead};
 
     #[test]
-    fn i2c_read_no_data_set() {
+    fn test_i2c_mock_write() {
+        let mut i2c = Mock::new();
+
+        i2c.expect(vec![Transaction::write(0xaa, vec![10u8, 12u8])]);
+
+        i2c.write(0xaa, &vec![10u8, 12u8]).unwrap();
+
+        i2c.done();
+    }
+
+    #[test]
+    fn test_i2c_mock_read() {
+        let mut i2c = Mock::new();
+
+        i2c.expect(vec![Transaction::read(0xaa, vec![1u8, 2u8])]);
+
+        let mut buff = vec![0u8; 2];
+        i2c.read(0xaa, &mut buff).unwrap();
+        assert_eq!(vec![1u8, 2u8], buff);
+
+        i2c.done();
+    }
+
+    #[test]
+    fn test_i2c_mock_write_read() {
+        let mut i2c = Mock::new();
+
+        i2c.expect(vec![Transaction::write_read(0xaa,
+            vec![1u8, 2u8],
+            vec![3u8, 4u8],
+        )]);
+
+        let v = vec![1u8, 2u8];
+        let mut buff = vec![0u8; 2];
+        i2c.write_read(0xaa, &v, &mut buff).unwrap();
+        assert_eq!(vec![3u8, 4u8], buff);
+
+        i2c.done();
+    }
+
+    #[test]
+    fn test_i2c_mock_multiple() {
         let mut i2c = Mock::new();
-        let mut buf = [0; 3];
-        i2c.read(0, &mut buf).unwrap();
-        assert_eq!(buf, [0; 3]);
+
+        i2c.expect(vec![
+            Transaction::write(0xaa, vec![1u8, 2u8]),
+            Transaction::read(0xbb, vec![3u8, 4u8]),
```

Comment:
Btw, I don't think you need to specify the `u8` suffix since Rust should be able to determine that from the `Transaction` type.